frontend/javascripts/oxalis/model/sagas/mapping_saga.ts

```diff
@@ -24,35 +43,108 @@ import {
   getMappingInfo,
   getVisibleSegmentationLayer,
 } from "oxalis/model/accessors/dataset_accessor";
-import type { ActiveMappingInfo, Mapping } from "oxalis/store";
+import type {
+  ActiveMappingInfo,
+  Mapping,
+  MappingType,
+  NumberLike,
+  NumberLikeMap,
+} from "oxalis/store";
 import ErrorHandling from "libs/error_handling";
 import { MAPPING_MESSAGE_KEY } from "oxalis/model/bucket_data_handling/mappings";
-import { api } from "oxalis/singletons";
-import { MappingStatusEnum } from "oxalis/constants";
-import { isMappingActivationAllowed } from "oxalis/model/accessors/volumetracing_accessor";
+import { Model } from "oxalis/singletons";
+import {
+  isMappingActivationAllowed,
+  hasEditableMapping,
+  getEditableMappingForVolumeTracingId,
+  needsLocalHdf5Mapping as getNeedsLocalHdf5Mapping,
+  getVolumeTracings,
+  getBucketRetrievalSourceFn,
+  BucketRetrievalSource,
+} from "oxalis/model/accessors/volumetracing_accessor";
 import Toast from "libs/toast";
 import { jsHsv2rgb } from "oxalis/shaders/utils.glsl";
 import { updateSegmentAction } from "../actions/volumetracing_actions";
+import { MappingStatusEnum } from "oxalis/constants";
+import DataCube from "../bucket_data_handling/data_cube";
+import { chainIterators, sleep } from "libs/utils";
+import { Action } from "../actions/actions";
+import { ActionPattern } from "redux-saga/effects";
+import { listenToStoreProperty } from "../helpers/listener_helpers";
+
 type APIMappings = Record<string, APIMapping>;
+type Container<T> = { value: T };
 
-const isAgglomerate = (mapping: ActiveMappingInfo) => {
-  if (!mapping) {
-    return false;
-  }
+const takeLatestMappingChange = (
+  oldActiveMappingByLayer: Container<Record<string, ActiveMappingInfo>>,
+  layerName: string,
+) => {
+  return fork(function* () {
+    let lastWatcherTask;
+    let lastBucketRetrievalSource;
+
+    const needsLocalMappingChangedChannel = createNeedsLocalMappingChangedChannel(layerName);
+    const getBucketRetrievalSourceForLayer = getBucketRetrievalSourceFn(layerName);
+    while (true) {
+      lastBucketRetrievalSource = yield* select((state) => getBucketRetrievalSourceForLayer(state));
+      const bucketRetrievalSource = yield* take(needsLocalMappingChangedChannel);
+
+      console.log("changed from", lastBucketRetrievalSource, "to", bucketRetrievalSource);
+
+      if (lastWatcherTask) {
+        console.log("Cancel old bucket watcher");
+        yield cancel(lastWatcherTask);
+        lastWatcherTask = null;
+      }
 
-  return mapping.mappingType === "HDF5";
+      // Changing between REQUESTED-WITH-MAPPING <> REQUESTED-WITHOUT-MAPPING
+      if (lastBucketRetrievalSource[0] !== bucketRetrievalSource[0]) {
+        yield* call(maybeReloadData, oldActiveMappingByLayer, { layerName }, true);
+      }
```

Comment:
> Kinda nitpicky but I have a suggestion to make it a little clearer that the reload is forces here (not requiring to read the arg names of maybeReloadData):

The code has changed in the meantime and the method is now simply `reloadData`. So, no "maybe" and no "force" parameter exists anymore :)

> And while I am at it: Is there a case where lastBucketRetrievalSource[0] !== bucketRetrievalSource[0] is false, but the lines below (starting the new bucket watcher), is actually needed? If not, could we add an else break statement here? -> this makes it much more clearer to read the while true loop here.
To ensure that the bucket watcher is not killed unncessesary, you could move the if of ll94 ff. below the else break statement

The entire code is only executed if the "bucket retrieval source" has changed. If `lastBucketRetrievalSource[0] === bucketRetrievalSource[0]` and needsLocalHdf5Mapping is true, the old watcher must be killed and then a new one must be spawned (since the computed mapping is different and all segment ids have to be send to the server again to map them properly). If needsLocalHdf5Mapping is false, the watcher is not needed anymore and still needs to be killed.
Therefore, I think, the code needs to stay as is :)